Extracting state logic into a reducer (#497)
reactjs/fr.react.dev: src/content/learn/extracting-state-logic-into-a-reducer.md

</Sandpack>

Vous n'aurez probablement pas besoin de le faire vous-même, mais c'est similaire à ce que fait React !

</DeepDive>

### Étape 3 : utilisez le réducteur depuis votre composant {/*step-3-use-the-reducer-from-your-component*/}

Pour finir, vous devez connecter le `tasksReducer` à votre composant. Importez le Hook `useReducer` depuis React :

```js
import { useReducer } from 'react';
```

Ensuite, vous pouvez remplacer le `useState` :

```js
const [tasks, setTasks] = useState(initialTasks);
```

par `useReducer` de cette façon :

```js
const [tasks, dispatch] = useReducer(tasksReducer, initialTasks);
```

> Le Hook `useReducer` est similaire à `useState` — vous devez lui passer un état initial and il renvoie une valeur d'état et un moyen de le redéfinir (dans ce cas, la fonction de *dispatch*). Toutefois, c'est un peu différent.

Reviewer: ```suggestion
Le Hook `useReducer` est similaire à `useState` : d'une part vous devez lui passer un état initial, d'autre part il renvoie une valeur d'état ainsi qu'un moyen de le redéfinir (en l'occurrence, la fonction de *dispatch*). Toutefois, des différences existent.
```

Pas mal de hiatus à lisser, du coup…

Le Hook `useReducer` prend deux arguments :

1. Une fonction de réduction.
2. Un état initial.

> et renvoie :

Reviewer: ```suggestion
Il renvoie :
```

> 1. Un valeur d'état.

Reviewer: ```suggestion
1. Une valeur d'état.
```

> 2. Un fonction *dispatch* (pour « *dispatcher* » les actions de l'utilisateur vers le réducteur).

Reviewer: ```suggestion
2. Une fonction *dispatch* (pour « *dispatcher* » les actions de l'utilisateur vers le réducteur).
```

1. A reducer function
2. An initial state
1. Une fonction de réduction.
2. Un état initial.

And it returns:
et renvoie :

1. A stateful value
2. A dispatch function (to "dispatch" user actions to the reducer)
1. Un valeur d'état.
2. Un fonction *dispatch* (pour « *dispatcher* » les actions de l'utilisateur vers le réducteur).

> Now it's fully wired up! Here, the reducer is declared at the bottom of the component file:

Reviewer: Argh, tu as purement et simplement viré la ligne !

      return tasks.filter((t) => t.id !== action.id);
    }
    default: {
      throw Error('Action inconnue : ' + action.type);
    }
  }
}

let nextId = 3;
const initialTasks = [

>   {id: 0, text: 'Visiter le musée de Kafka', done: true},

Reviewer: Bon en gros change partout, idem pour la photo du mur John Lennon…

            });
          }}
        />
        <button onClick={() => setIsEditing(false)}>Enregistrer</button>
      </>
    );
  } else {
    taskContent = (
      <>
        {task.text}

>         <button onClick={() => setIsEditing(true)}>Éditer</button>

Reviewer: ```suggestion
        <button onClick={() => setIsEditing(true)}>Modifier</button>
```

</Sandpack>

La logique des composants peut être plus simple à lire quand vous séparez les responsabilités de cette façon. Maintenant les gestionnaires d'événements spécifient seulement _ce qui s'est passé_ en *dispatchant* les actions, et la fonction de réduction détermine _comment l'état se met à jour_ en réponse à celles-ci.

## Comparaison de `useState` et `useReducer` {/*comparing-usestate-and-usereducer*/}

Les réducteurs ne sont pas sans inconvénients ! Voici quelques éléments de comparaison :

> - **Taille du code** : avec un `useState`, vous devez généralement écrire moins de code au début. Avec `useReducer`, vous devez écrire à la fois la fonction de réduction _et_ les actions de *dispatch*. Cependant, `useReducer` peut aider à réduire le code si de nombreux gestionnaires d'événements modifient l'état d'une façon similaire.

Reviewer: ```suggestion
- **Taille du code** : avec un `useState`, vous devez généralement écrire moins de code au début. Avec `useReducer`, vous devez écrire à la fois la fonction de réduction _et_ *dispatcher* les actions. Cependant, `useReducer` peut aider à réduire le code si plusieurs gestionnaires d'événements modifient l'état de façon similaire.
```

> - **Lisibilité** : `useState` est très facile à lire lorsque les mises à jour d'état sont simples. Quand ça se complique, elles peuvent gonfler le code de votre composant et le rendre difficile à analyser. Dans ce cas, `useReducer` vous permet de séparer proprement le _comment_ de la logique de _ce qui est arrivé_ des gestionnaires d'événements.

Reviewer: ```suggestion
- **Lisibilité** : `useState` est très facile à lire lorsque les mises à jour d'état sont simples. Quand ça se complique, elles peuvent gonfler le code de votre composant et le rendre difficile à analyser. Dans ce cas, `useReducer` vous permet de séparer proprement le _comment_ de la logique du _ce qui est arrivé_ des gestionnaires d'événements.
```

> - **Débogage** : quand vous avez un bug avec un `useState`, il peut être difficile de dire _où_ l'état a été définit incorrectement et _pourquoi_. Avec un `useReducer`, vous pouvez ajouter des messages dans la console dans votre réducteur pour voir chaque mise à jour d'état et _pourquoi_ elles ont lieu (en rapport à quelle `action`). Si chaque `action` est correcte, vous saurez que le problème se trouve dans la logique de réduction elle-même. En revanche, vous devez parcourir plus de code qu'avec `useState`.

Reviewer: ```suggestion
- **Débogage** : quand vous avez un bug avec un `useState`, il peut être difficile de dire _où_ l'état a été mal défini et _pourquoi_. Avec un `useReducer`, vous pouvez ajouter des messages dans la console depuis votre réducteur pour voir chaque mise à jour d'état et _pourquoi_ elles ont lieu (en rapport à quelle `action`). Si chaque `action` est correcte, vous saurez que le problème se trouve dans la logique de réduction elle-même. En revanche, vous devez parcourir plus de code qu'avec `useState`.
```

Attention à ton "définit" qui n'est pas un participe passé, tu fais fréquemment l'erreur (c'est la 3e personne du présent de l'indicatif).

> - **Tests** : un réducteur est une fonction pure qui ne dépend pas de votre composant. Ça signifie que vous pouvez l'exporter et la tester en isolation. Bien qu'il soit généralement préférable de tester des composants dans un environnement plus réaliste, pour une logique de mise à jour d'état plus complexe, il peut être utile de vérifier que votre réducteur renvoie un état particulier pour un état initial et une action particuliers.

Reviewer: ```suggestion
- **Tests** : un réducteur est une fonction pure qui ne dépend pas de votre composant. Ça signifie que vous pouvez l'exporter et la tester en isolation. Bien qu'il soit généralement préférable de tester des composants dans un environnement plus réaliste, pour une logique de mise à jour d'état plus complexe, il peut être utile de vérifier que votre réducteur renvoie un état spécifique pour un état initial et une action particuliers.
```

- **Préférence personnelle** : certaines personnes aiment les réducteurs, d'autres non. Ce n'est pas grave. C'est une question de préférence. Vous pouvez toujours convertir un `useState` en un `useReducer` et inversement : ils sont équivalents !

> Nous recommandons d'utiliser un réducteur si vous rencontrez souvent des bugs à cause de mauvaises mises à jour d'état dans certains composants et que vous souhaitez introduire plus de structure dans son code. Vous n'êtes pas obligé d'utiliser les réducteurs pour tout : n'hésitez pas à les mélanger ! Vous pouvez aussi utiliser `useState` et `useReducer` dans le même composant.

Reviewer: ```suggestion
Nous recommandons d'utiliser un réducteur si vous rencontrez souvent des bugs à cause de mauvaises mises à jour d'état dans un composant et que vous souhaitez introduire plus de structure dans son code. Vous n'êtes pas obligé·e d'utiliser les réducteurs pour tout : n'hésitez pas à mélanger les approches ! Vous pouvez aussi utiliser `useState` et `useReducer` dans le même composant.
```

## Écrire les réducteurs correctement {/*writing-reducers-well*/}

Gardez ces deux points à l'esprit quand vous écrivez des réducteurs :

> - **Les réducteurs doivent être purs**. Tout comme les [fonctions de mise à jour d'état](/learn/queueing-a-series-of-state-updates), les réducteurs sont exécutés pendant le rendu (les actions sont mises en attente jusqu'au rendu suivant) ! Ça signifie que les réducteurs [doivent être purs](/learn/keeping-components-pure) — les mêmes entrées résultent toujours dans les mêmes sorties. Ils ne doivent pas envoyer de requêtes, planifier des *timeouts* ou effectuer des effets secondaires (des opérations qui impactent des éléments extérieurs au composant). Ils doivent mettre à jour des [objets](/learn/updating-objects-in-state) et [des tableaux](/learn/updating-arrays-in-state) sans modifications.

Reviewer: ```suggestion
- **Les réducteurs doivent être purs.** Tout comme les [fonctions de mise à jour d'état](/learn/queueing-a-series-of-state-updates), les réducteurs sont exécutés pendant le rendu! ! (Les actions sont mises en attente jusqu'au rendu suivant.) Ça signifie que les réducteurs [doivent être purs](/learn/keeping-components-pure) — les mêmes entrées produisent toujours les mêmes sorties. Ils ne doivent pas envoyer de requêtes, planifier des *timers* ou traiter des effets secondaires (des opérations qui impactent des entités extérieures au composant). Ils doivent mettre à jour des [objets](/learn/updating-objects-in-state) et [des tableaux](/learn/updating-arrays-in-state) en respectant l'immutabilité.
```

> - **Chaque action décrit une interaction utilisateur unique, même si ça entraîne plusieurs modifications des données**. Par exemple, si un utilisateur appuie sur le bouton « Réinitialiser » d'un formulaire comportant cinq champs gérés par un réducteur, il sera plus logique de *dispatcher* une seule action `reset_form` plutôt que cinq actions `set_field` distinctes. Si vous enregistrez chaque action dans un réducteur, cet enregistrement doit être suffisament clair pour vous permettre de reconstruire les interactions ou les réponses qui se sont produites dans le bon ordre. Ça aide au débogage !

Reviewer: ```suggestion
- **Chaque action décrit une interaction utilisateur unique, même si ça entraîne plusieurs modifications des données.** Par exemple, si l'utilisateur appuie sur le bouton « Réinitialiser » d'un formulaire comportant cinq champs gérés par un réducteur, il sera plus logique de *dispatcher* une seule action `reset_form` plutôt que cinq actions `set_field` distinctes. Si vous journalisez chaque action d'un réducteur, ce journal doit être suffisamment clair pour vous permettre de reconstruire l'ordre et la nature des interactions et de leurs traitements. Ça facilite le débogage !
```

## Écrire des réducteurs concis avec Immer {/*writing-concise-reducers-with-immer*/}

> Tout comme [mettre à jour des objets](/learn/updating-objects-in-state#write-concise-update-logic-with-immer) et [des tableaux](/learn/updating-arrays-in-state#write-concise-update-logic-with-immer) dans un état ordinaire, vous pouvez utiliser la bibliothèque Immer pour rendre les réducteurs plus concis. Ici, [`useImmerReducer`](https://github.com/immerjs/use-immer#useimmerreducer) vous permet de modifier l'état avec une instruction `push` ou `arr[i] =` :

Reviewer: ```suggestion
Comme  pour la [mise à jour des objets](/learn/updating-objects-in-state#write-concise-update-logic-with-immer) et [des tableaux](/learn/updating-arrays-in-state#write-concise-update-logic-with-immer) dans un état ordinaire, vous pouvez utiliser la bibliothèque Immer pour rendre les réducteurs plus concis. Ici, [`useImmerReducer`](https://github.com/immerjs/use-immer#useimmerreducer) vous permet de modifier l'état avec un appel à `push` ou encore une affectation `arr[i] =` :
```

</Sandpack>

Les réducteurs doivent être purs, donc ils ne doivent pas modifier l'état. Cependant, Immer fournit un objet spécial `draft` qu'il est possible de modifier. Sous le capot, Immer créera une copie de votre état avec les changements que vous avez appliqué sur le `draft`. C'est pourquoi les réducteurs gérés par `useImmerReducer` peuvent modifier leur premier argument et n'ont pas besoin de renvoyer l'état.

<Recap>

- Pour convertir `useState` vers `useReducer` :
  1. *Dispatchez* les actions depuis des gestionnaires d'événements.
  2. Écrivez une fonction de réduction qui renvoie le prochain état en fonction d'un état et d'une action donnés.
  3. Remplacez `useState` par `useReducer`.
- Les réducteurs vous obligent à écrire un peu plus de code, mais ils facilitent le débogage et les tests.
- Les réducteurs doivent être purs.
- Chaque action décrit une interaction utilisateur unique.

> - Utilisez Immer si vous souhaitez écrire des réducteurs avec un style de modification.

Reviewer: ```suggestion
- Utilisez Immer si vous souhaitez écrire des réducteurs en modifiant directement l'état entrant.
```

</Recap>

<Challenges>

#### *Dispatcher* des actions depuis des gestionnaires d'événements {/*dispatch-actions-from-event-handlers*/}

Pour l'instant, les gestionnaires d'événements dans `ContactList.js` et `Chat.js` contiennent des commentaires `// TODO`. C'est pour ça que taper dans le champ de saisie ne marche pas, et cliquer sur les boutons ne change pas le destinataire sélectionné.

> Remplacez ces deux commentaires `// TODO` avec le code faisant le `dispatch` des actions correspondantes. Pour voir la forme attendue et le type des actions, vérifiez le réducteur dans `messengerReducer.js`. Il est déjà écrit, vous n'avez donc pas besoin de le changer. Vous devez seulement *dispatcher* les actions dans `ContactList.js` et `Chat.js`.

Reviewer: ```suggestion
Remplacez ces deux commentaires `// TODO` par du code qui `dispatch` les actions correspondantes. Pour connaître la forme attendue et le type des actions, allez voir le réducteur dans `messengerReducer.js`. Il est déjà écrit, vous n'avez donc pas besoin de le changer. Vous devez seulement *dispatcher* les actions dans `ContactList.js` et `Chat.js`.
```

<Hint>

La fonction `dispatch` est déjà disponible dans les deux composants parce qu'elle a été passée comme une prop. Vous devez donc appeler `dispatch`avec l'objet d'action correspondant.

> Pour vérifier la forme de l'objet d'action, vous pouvez regarder le réducteur et voir quels champs d'`action` il s'attend à avoir. Par exemple, le cas de l'action `changed_selection` du réducteur ressemble à ceci :

Reviewer: ```suggestion
Pour vérifier la forme de l'objet d'action, vous pouvez examiner le réducteur et voir à quels champs d'`action` il s'attend. Par exemple, le cas de l'action `changed_selection` du réducteur ressemble à ceci :
```

<Solution>

> Avec le code du réducteur, vous pouvez déduire que les actions doivent ressembler à ceci :

Reviewer: ```suggestion
Sur la base du code du réducteur, vous pouvez déduire que les actions doivent ressembler à ceci :
```

Avec le code du réducteur, vous pouvez déduire que les actions doivent ressembler à ceci :

```js
// When the user presses "Alice"
// Quand l’utilisateur choisit "Alice"
dispatch({
  type: 'changed_selection',
  contactId: 1,
});

// When user types "Hello!"
// Quand l’utilisateur tape « Bonjour ! »
dispatch({
  type: 'edited_message',
  message: 'Hello!',
  message: 'Bonjour !',
});
```

> Here is the example updated to dispatch the corresponding messages:

Reviewer: Ben ! Où est la trad ?!

Author: Je vais les remettre. J'avais fait une fausse manipulation dans l'IDE, plusieurs lignes avaient été perdues. Je pensais les avoir remises, mais j'en ai visiblement oublié 2...

</Solution>

#### Vider la saisie après l'envoi d'un message {/*clear-the-input-on-sending-a-message*/}

Pour l'instant, appuyer sur « Envoyer » ne fait rien du tout. Ajoutez un gestionnaire d'événement au bouton « Envoyer » qui va :

> 1. Afficher une `alert` avec l'email du destinataire et le message.

Reviewer: ```suggestion
1. Afficher une `alert` avec l'e-mail du destinataire et le message.
```

</Sandpack>

> Ça fonctionne et le champ de saisie se vide quand vous appuyez sur « Envoyer ».

Reviewer: ```suggestion
Ça fonctionne et le champ de saisie est vidé quand vous appuyez sur « Envoyer ».
```

</Sandpack>

> Le comportement ainsi obtenu est le même. Cependant, gardez à l'esprit que les types d'action doivent idéalement décrire « ce qu'a fait l'utilisateur » plutôt que « la façon dont vous voulez que l'état change ». Ça facilite l'ajout de fonctionnalités par la suite.

Reviewer: ```suggestion
Le comportement visible ne change pas. Cependant, gardez à l'esprit que les types d'action doivent idéalement décrire « ce qu'a fait l'utilisateur » plutôt que « la façon dont vous voulez que l'état change ». Ça facilite l'ajout de fonctionnalités par la suite.
```

> Avec l'une ou l'autre des solutions, il est important que vous ne placiez *pas* l'`alert` dans un réducteur. Le réducteur doit être une fonction pure — il doit se limiter au calcul du prochain état. Il ne doit rien « faire », y compris afficher des messages à l'utilisateur. Ça doit se produire dans le gestionnaire d'événement (pour détecter ce genre d'erreur, React appelera vos réducteurs plusieurs fois dans le mode Strict. De cette façon, si vous placez une alerte dans un réducteur, vous la verrez deux fois).

Reviewer: ```suggestion
Quelle que soit l'approche retenue, il est important que vous ne placiez *pas* l'`alert` dans un réducteur. Le réducteur doit être une fonction pure : il doit se limiter au calcul du prochain état. Il ne doit rien « faire », y compris afficher des messages à l'utilisateur. Ça doit être fait dans le gestionnaire d'événement (pour détecter ce genre d'erreur, React en mode strict appellera vos réducteurs plusieurs fois. Du coup, si vous placez une alerte dans un réducteur, vous la verrez deux fois).
```

</Solution>

#### Restaurer les valeurs de saisie au changement d'onglet {/*restore-input-values-when-switching-between-tabs*/}

Dans cet exemple, changer de destinataires vide toujours le champ de saisie :

```js
case 'changed_selection': {
  return {
    ...state,
    selectedId: action.contactId,
    message: '' // Vide la saisie
  };
```

> En effet, vous ne souhaitez pas partager un seul brouillon de message entre plusieurs destinataires. Néanmoins, il serait préférable que votre appli « se souvienne » des brouillons pour chaque contact séparément, les restaurant quand vous changez de contact.

Reviewer: ```suggestion
En effet, vous ne souhaitez pas partager un seul brouillon de message entre plusieurs destinataires. Néanmoins, il serait préférable que votre appli « se souvienne » des brouillons pour chaque contact séparément, et les restaure quand vous changez de contact.
```

Votre tâche consiste à modifier la façon dont l'état est structuré afin de mémoriser le brouillon _pour chaque contact_. Vous aurez besoin d'appliquer quelques changements au réducteur, à l'état initial ainsi qu'aux composants.

<Hint>

Vous pouvez changer la structure de votre état ainsi :

```js
export const initialState = {
  selectedId: 0,
  messages: {

>     0: 'Bonjour, Taylor', // Brouillon pour contactId = 0

Reviewer: Pour info, partout dans les docs on remplace Taylor (Swift) par Clara (Luciani), qu'on a choisie comme "équivalent français" de Taylor Swift en termes de pop culture 😉. Rien de bien grave, mais autant préserver la cohérence !

Author: OK, ça me va, ma fille s'appelle Clara 😉

    1: 'Bonjour, Alice', // Brouillon pour contactId = 1
  },
};
```

> La syntaxe `[key]: value` des [propriétés calculées](https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Operators/Object_initializer#noms_de_propri%C3%A9t%C3%A9s_calcul%C3%A9s) peut vous aider à mettre à jour l'objet `messages` :

Reviewer: ```suggestion
La syntaxe `[key]: value` des [noms de propriétés calculés](https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Operators/Object_initializer#noms_de_propri%C3%A9t%C3%A9s_calcul%C3%A9s) peut vous aider à mettre à jour l'objet `messages` :
```

</Sandpack>

> Notez que vous n'avez pas eu besoin de modifier les gestionnaires d'événements pour implémenter ce comportement différent. Sans réducteur, vous auriez dû changer chaque gestionnaire d'événement qui met à jour l'état.

Reviewer: ```suggestion
Notez que vous n'avez pas eu besoin de modifier les gestionnaires d'événements pour implémenter ce nouveau comportement. Sans réducteur, vous auriez dû changer chaque gestionnaire d'événement qui met à jour l'état.
```

</Solution>

#### Implémenter `useReducer` de zéro {/*implement-usereducer-from-scratch*/}

Dans les exemples précédents, vous avez importé le Hook `useReducer` de React. Cette fois, vous allez implémenter _le Hook `useReducer` vous-même_ ! Voici un bout de code pour démarrer. Ça ne devrait pas vous prendre plus de 10 lignes de code.

Pour tester vos modifications, essayez de taper dans le champ de saisie ou choisissez un contact.

<Hint>

> Voici une esquisse plus détaillée de l'implémentation :

Reviewer: Ah ah, pas mal "esquisse" 😁

}
```

> Souvenez-vous qu'une fonction de réduction prend deux arguments — l'état courant et l'objet d'action — et renvoie le prochain état. Que doit-en faire votre implémentation de `dispatch` ?

Reviewer: ```suggestion
Souvenez-vous qu'une fonction de réduction prend deux arguments — l'état courant et l'objet d'action — et renvoie le prochain état. Que doit en faire votre implémentation de `dispatch` ?
```

<Solution>

> *Dispatcher* une action appelle un réducteur avec l'état actuel et l'action, puis conserve le résultat comme le prochain état. Voici à quoi ça ressemble dans le code :

Reviewer: ```suggestion
*Dispatcher* une action appelle un réducteur avec l'état actuel et l'action, puis conserve le résultat comme prochain état. Voici à quoi ça ressemble dans le code :
```

</Sandpack>

Bien que ça n'ait pas d'importance dans la plupart des cas, une implémentation un peu plus précise ressemble à ça :

```js
function dispatch(action) {
  setState((s) => reducer(s, action));
}
```

> En effet, les actions *dispatchées* sont mises en attente jusqu'au prochain rendu, [à l'image des fonctions de mise à jour](/learn/queueing-a-series-of-state-updates).

Reviewer: ```suggestion
En effet, les actions *dispatchées* sont mises en file d'attente jusqu'au prochain rendu, [tout comme les fonctions de mise à jour](/learn/queueing-a-series-of-state-updates).
```
